### PR TITLE
vector_layers

### PR DESCRIPTION
**Pull request by mapsam**
This documents `vector_layers` as a required key in TileJSON (for sets of vector tiles). Per discussion in #14 this aims to "describe" the layers and their possible attributes while stopping short of describing geometry.

The language is adapted from the [vector_layers section](https://github.com/mapbox/mbtiles-spec/blob/master/1.3/spec.md#vector_layers) in the 1.3 version of the mbtiles-spec, which is the first place to describe this data structure. The key differences between this and the MBTiles specification description is the tilejson specification does not limit the values in key-value pairs of the `fields` attributes to be `Number`, `String`, or `Boolean` and simply suggests this as a space to describe the key with a string. Per discussion with @GretaCB and @ericfischer, this may warrant an update to the MBTiles specification to be less strict.

*Note: we'll be opening a ticket to document possibly renaming `id` to `name` in version 4.0, as suggested by @pnorman in https://github.com/mapbox/tilejson-spec/issues/14#issuecomment-251670418*

refs: https://github.com/mapbox/tilejson-spec/issues/35

### CODE REVIEW

File: 3.0/README.md [hunk 1]
```diff
@@ -154,6 +154,55 @@ An array of tile endpoints. {z}, {x} and {y}, if present, are replaced with the
...
+
+Each layer object MAY also contain the following key-value pair:
+
+* `description` (string): A human-readable description of the layer's contents.
+* `minzoom` (number): The lowest zoom level whose tiles this layer appears in. MUST be greater than or equal to the tileset's `minzoom`.
+* `maxzoom` (number): The highest zoom level whose tiles this layer appears in. MUST be less than or equal to the tileset's `maxzoom`.
+
+These keys are used to describe the situation where different sets of vector layers appear in different zoom levels of the same tileset, for example in a case where a "minor roads" layer is only present at high zoom levels.
+
+Implementations MUST treat unknown key-value pairs as if they weren't present. However, implementations MUST expose unknown key/values in their API so that API users can optionally handle these keys.
```
Comment (mapsam) [inline 3.0/README.md]: @GretaCB I've noticed some implementations (mapbox streets tilejson for example) include other keys here. I'm thinking we can document that this is acceptable with the same language from the Structure section https://github.com/mapbox/tilejson-spec/blob/220b1d221a53809fe9d955c611db11d219e99911/3.0/README.md#3-structure. Duplicating the text feels a little redundant, so I'm wondering if we could update the Structure introduction to say something like:

> Implementations MUST treat unknown keys as if they weren't present. **This is true at any depth of the JSON**. However, implementations MUST expose unknown key/values in their API so that API users can optionally handle these keys.

File: 3.0/README.md [hunk 2]
```diff
@@ -154,6 +154,55 @@ An array of tile endpoints. {z}, {x} and {y}, if present, are replaced with the
 ```
 
 ## 3.15 `vector_layers`
+
+REQUIRED if describing vector tiles.
```
Comment (mapsam) [inline 3.0/README.md]: We're saying this is required, but how do we effectively cover non vector-tile cases that cannot be described in this format?

**Comment by mapsam** [comment]
I opened a ticket in the mbtiles specification repo to suggest relaxing the requirements for `String`, `Number`, or `Boolean`, which we can think about once the tilejson spec 3.0 has been released: https://github.com/mapbox/mbtiles-spec/issues/53

File: 3.0/README.md [hunk 3]
```diff
@@ -154,6 +154,55 @@ An array of tile endpoints. {z}, {x} and {y}, if present, are replaced with the
 ```
 
 ## 3.15 `vector_layers`
```
Comment (pnorman) [inline 3.0/README.md]: I'd add subsections, defining each of the sub-items.

File: 3.0/README.md [hunk 4]
```diff
@@ -154,6 +154,55 @@ An array of tile endpoints. {z}, {x} and {y}, if present, are replaced with the
...
+
+REQUIRED. (if describing vector tiles)
+
+A JSON object whose value is an array of JSON objects. Each of those JSON objects describes one layer of vector tile data, and MUST contain the following key-value pairs:
+
+* `id` (string): The layer ID, which is referred to as the `name` of the layer in the [Mapbox Vector Tile spec](https://github.com/mapbox/vector-tile-spec/tree/master/2.1#41-layers).
+* `fields` (object): A JSON object whose keys and values are the names and descriptions of attributes available in this layer. Each value (description) MUST be a string. The values MAY describe the underlying data type, such as `String`, `Number`, or `Boolean`. Attributes whose type varies between features MAY be listed as `"String"`.
```
Comment (pnorman) [inline 3.0/README.md]: It's not really clear from this that the values are free-form text, and no special meaning should be given to an attempt to describe the underlying data type, since "Number" could be a description or a type, and different tile formats might have different allowable types

File: 3.0/README.md [hunk 5]
```diff
@@ -154,6 +154,55 @@ An array of tile endpoints. {z}, {x} and {y}, if present, are replaced with the
 ```
 
 ## 3.15 `vector_layers`
+
+REQUIRED. (if describing vector tiles)
+
+A JSON object whose value is an array of JSON objects. Each of those JSON objects describes one layer of vector tile data, and MUST contain the following key-value pairs:
+
+* `id` (string): The layer ID, which is referred to as the `name` of the layer in the [Mapbox Vector Tile spec](https://github.com/mapbox/vector-tile-spec/tree/master/2.1#41-layers).
```
Comment (pnorman) [inline 3.0/README.md]: It's not clear to me what the reference to the MVT spec is.

File: 3.0/README.md (hunk 1 above)
Comment (pnorman) [reply 3.0/README.md]: I don't think this section is needed at all, given the text elsewhere in the spec.

File: 3.0/README.md (hunk 5 above)
Comment (pnorman) [inline 3.0/README.md]: It should be explicit that there is no requirement the id is unique, but some formats may prohibit overlapping id and zoom.

File: 3.0/README.md [hunk 6]
```diff
@@ -154,6 +154,55 @@ An array of tile endpoints. {z}, {x} and {y}, if present, are replaced with the
 ```
 
 ## 3.15 `vector_layers`
+
+REQUIRED. (if describing vector tiles)
```
Comment (pnorman) [inline 3.0/README.md]: I'm not sure what this means.

File: 3.0/README.md [hunk 7]
```diff
@@ -154,6 +154,55 @@ An array of tile endpoints. {z}, {x} and {y}, if present, are replaced with the
...
+        "iso": "ISO 3166-1 Alpha-2 code",
+        "name": "English name of the country",
+        "name_ar": "Arabic name of the country"
+      }
+    }
+  ]
+}
+```
+
+
```
Comment (pnorman) [inline 3.0/README.md]: Stray newline

File: 3.0/README.md [hunk 8]
```diff
@@ -116,7 +116,7 @@ A name describing the tileset. The name can contain any legal character. Impleme
 
 ## 3.11 `scheme`
 
-OPTIONAL. Default: "xyz". 
+OPTIONAL. Default: "xyz".
```
Comment (pnorman) [inline 3.0/README.md]: This whitespace change should go in a different commit because it's unrelated.

File: 3.0/README.md [hunk 9]
```diff
@@ -141,7 +141,7 @@ A semver.org style version number. Describes the version of the TileJSON spec th
 
 ## 3.14 `tiles`
 
-REQUIRED. 
+REQUIRED.
```
Comment (pnorman) [inline 3.0/README.md]: This whitespace change should go in a different commit because it's unrelated.

**Comment by pnorman** [comment]
> Note: we'll be opening a ticket to document possibly renaming id to name in version 4.0, as suggested by @pnorman in #14 (comment)

If we want to call it `name`, let's do it now, rather than require a semver major change to do it.

File: 3.0/README.md (hunk 4 above)
Comment (mapsam) [reply 3.0/README.md]: Agree, I think we should remove the concept of `Number` `String` `Boolean` entirely and just say this is string value that should describe the property. The explicit options are leftover from the mbtiles-spec, which is likely going to update once we ship v3.

File: 3.0/README.md (hunk 1 above)
Comment (GretaCB) [reply 3.0/README.md]: @mapsam what if we link to the Structure section to avoid needing to update text in both places, but also keep a reference in both places to avoid it being overlooked?

File: 3.0/README.md (hunk 5 above)
Comment (mapsam) [reply 3.0/README.md]: Per discussion in #14 it is to help connect what this value refers to according to the vector tile layer concept. The MVT spec is used as an example. If we remove this reference entirely, it could be confusing since it’s nice to point to how vector tile specifications refer to layer names. I think we can remove the reference in v4 once we rename the property to `name`. 

Basically, if we're considering matching the `id` field to `name` to match the MVT specification, then the reference link to the MVT specification feels relevant.

Comment (mapsam) [reply 3.0/README.md]: Agreed, the enforcement of this is vague. 

cc @flippmoke @GretaCB

**Comment by mapsam** [comment]
Thanks for the comments @pnorman! A couple more responses inline here:

> If we want to call it name, let's do it now, rather than require a semver major change to do it.

The goal of 3.x is to solidify the current common practice and reduce the need for architectural changes for implementations and developers. It’s not perfect, but it is what it is. 4.x can introduce downstream architectural changes and will be a more appropriate space for this.

> The language identifying the required items should mirror that of the rest of the spec. I think this will be easier to do with subsections since the structure will then be the same too.

Great idea - definitely agree. I'll rework this section to have appropriate subsections.

File: 3.0/README.md (hunk 1 above)
Comment (mapsam) [reply 3.0/README.md]: Sounds good to me @GretaCB - will work an update!

**Comment by mapsam** [comment]
Subsections have been added and a note about how v4 can rename `id` to `name`.

@GretaCB ready for another review here!

**Comment by GretaCB** [comment]
Merging 👍
